File: Swift/Topological Sort/1005.swift
import Foundation

struct Queue {
    var In = [Int]()
    var Out = [Int]()
    
    mutating func append(_ val:Int) {
        In.append(val)
    }
    
    mutating func pop() -> Int {
        if Out == [] {
            Out = In.reversed()
            In = []
        }
        return Out.popLast()!
    }
}

var T = Int(readLine()!)!
for _ in 1...T {
    var input = readLine()!.split(separator: " ").map { Int($0)! }
    let N = input[0], K = input[1]
    let timeArr: [Int] = readLine()!.split(separator: " ").map { Int($0)! }
    var addedTime = Array(repeating: 0, count: N)
    var In = Array(repeating: 0, count: N+1)
    var Out: [[Int]] = Array(repeating: [], count: N+1)
    for _ in 1...K {
        input = readLine()!.split(separator: " ").map { Int($0)! }; let A = input[0]; let B = input[1];
        Out[A].append(B)
        In[B] += 1
    }
    let W = Int(readLine()!)!
    var que = Queue()
    for i in 1...N {
        if In[i] == 0 {
            que.append(i)
            addedTime[i-1] = timeArr[i-1]
        }
    }
    while( que.In != [] || que.Out != [] ) {
        let now = que.pop()
        for out in Out[now] {
            In[out] -= 1
            addedTime[out-1] = max(addedTime[now-1]+timeArr[out-1], addedTime[out-1])
            if In[out] == 0 {
                que.append(out)
            }
        }
    }
    print(addedTime[W-1])
}

File: JavaScript/Topological Sort/1005.js
class Queue {
    constructor() {
      this.s = {};
      this.head = 0;
      this.tail = 0;
    }
    add(val) {
      this.s[this.tail++] = val;
    }
    popfront() {
      const out = this.s[this.head];
      delete this.s[this.head++];
      return out;
    }
    isEmpty() {
      return this.head === this.tail;
    }
}
  
const input = require("fs").readFileSync("/dev/stdin").toString().trim().split("\n");
  
let idx = 0;
const T = Number(input[idx++]);
  
for (let t = 0; t < T; t++) {
    const [N, K] = input[idx++].split(" ").map(Number);
    const buildTime = input[idx++].split(" ").map(Number);
  
    const outE = Array.from({ length: N }, () => new Set());
    const inE = Array.from({ length: N }, () => new Set());
  
    for (let i = 0; i < K; i++) {
      const [start, end] = input[idx++].split(" ").map(Number);
      outE[start - 1].add(end - 1);
      inE[end - 1].add(start - 1);
    }
  
    const W = Number(input[idx++]) - 1;
  
    const que = new Queue();
    const times = Array(N).fill(0);
  
    for (let i = 0; i < N; i++) {
      if (inE[i].size === 0) {
        que.add(i);
        times[i] = buildTime[i];
      }
    }
  
    while (!que.isEmpty()) {
      const now = que.popfront();
      for (let out of outE[now]) {
        inE[out].delete(now);
        times[out] = Math.max(times[out], times[now] + buildTime[out]);
        if (inE[out].size === 0) {
          que.add(out);
        }
      }
    }
  
    console.log(times[W]);
}